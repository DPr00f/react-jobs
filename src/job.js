/* @flow */

import React, { Component, PropTypes } from 'react';
import { getDisplayName, isPromise } from './utils';
import type { ProviderContext } from './sharedTypes';

type Work = (props : Object) => any;

type State = {
  resolved?: boolean,
  inProgress: boolean,
  result?: any,
  error?: any,
  executingWork? : Promise<any>,
};

type Props = Object;

export default function job(work : Work) {
  if (typeof work !== 'function') {
    throw new Error('You must provide a function to a react-jobs job declaration.');
  }

  // Oh I love you closures!
  let jobID = null;

  return function WrapComponentWithJob(WrappedComponent : Function) {
    class ComponentWithJob extends Component {
      props: Props;
      state: State;

      constructor(props : Props, context : ?ProviderContext) {
        super(props);
        if (context && context.reactJobs) {
          if (!jobID) {
            jobID = context.reactJobs.nextJobID();
          } else {
            const jobResults = context.reactJobs.getJobResults(jobID);
            if (jobResults) {
              this.state = Object.assign(
                {},
                jobResults,
                { resolved: true },
              );
            }
          }
        }
        if (!this.state) {
          this.state = { inProgress: false };
        }
      }

      componentWillMount() {
        if (this.state.resolved) {
          return;
        }

        const context : ProviderContext = this.context;
        const x = work(this.props);

        if (isPromise(x)) {
          const executingWork = x
            .then((result) => { this.setState({ inProgress: false, result }); return result; })
            .catch((error) => {
              console.warn('Job failed:\n', error); // eslint-disable-line no-console
              this.setState({ inProgress: false, error });
            })
            .then(() => {
              if (jobID) {
                context.reactJobs.registerJobResults(jobID, this.state);
              }
            });

          // Asynchronous result.
          this.setState({ inProgress: true, executingWork });
        } else {
          // Synchronous result.
          this.setState({ result: x });
        }
      }

      getExecutingJob() {
        if (!this.state.executingWork) {
          return undefined;
        }
        return this.state.executingWork;
      }

      render() {
        const { inProgress, result, error } = this.state;
        const jobProp = { inProgress, result, error };
        return <WrappedComponent {...this.props} job={jobProp} />;
      }
    }
    ComponentWithJob.displayName = `${getDisplayName(WrappedComponent)}WithJob`;
    ComponentWithJob.contextTypes = { reactJobs: PropTypes.object };
    return ComponentWithJob;
  };
}
