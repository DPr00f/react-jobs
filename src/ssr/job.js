/* @flow */
// @see https://github.com/facebook/react/issues/1137

import React, { PropTypes } from 'react';
import { getDisplayName } from '../utils';
import clientOnlyJob from '../job';
import type { ReactJobsClientContext, ReactJobsServerContext } from './types';

type Context = {
  reactJobsClient: ReactJobsClientContext,
  reactJobsServer?: ReactJobsServerContext,
};

const getInitialState = (context, jobID) => {
  const { reactJobsClient, reactJobsServer } = context;

  if (reactJobsServer) {
    // Running on the server. If the jobs have been executed by the runJobs
    // helper then we will get back a result here.
    const jobState = reactJobsServer.getJobState(jobID);
    if (jobState) {
      return jobState;
    }
  } else {
    // Otherwise we have to be on the browser. Check if there is any
    // rehydration state available for this job.
    const ssrRehydrateState = reactJobsClient.popJobRehydrationForSRR(jobID);
    if (ssrRehydrateState) {
      return ssrRehydrateState;
    }
  }

  return undefined;
};

const job = (work : any) => (Component : Function) => {
  let jobIDHandle = null;

  function withSSRBehaviour(WrappedComponent) {
    const ComponentWithJobID = (props : Object, context : Context) => {
      const { reactJobsClient, reactJobsServer } = context;

      // Every SSR job should have the client attached in order to provide
      // the identifiers for the components.
      if (!reactJobsClient) {
        throw new Error('You are using an "ssr/job" without having wrapped your component render with the "runJobs" or "rehydrateJobs" helpers.  Please refer to the docs for more details.');
      }

      // Establish a unique identifier for this job instance.
      const getJobID = () => {
        if (!jobIDHandle) {
          jobIDHandle = reactJobsClient.nextJobID();
        }
        return jobIDHandle;
      };
      const jobID = getJobID();

      // Determine if we have an state from server side processing or
      // client side rehdration.
      const initialState = getInitialState(context, jobID);

      // When on a server we want to call back and register the result of the
      // work so that the server can hydrate the app appropriately as well
      // as provide the initial state to be sent to the browser.
      const onJobProcessed = reactJobsServer
        ? jobsState => reactJobsServer.registerJobState(jobID, jobsState)
        : null;

      return (
        <WrappedComponent
          {...props}
          jobInitState={initialState}
          onJobProcessed={onJobProcessed}
        />
      );
    };
    ComponentWithJobID.displayName = `${getDisplayName(WrappedComponent)}WithJobID`;
    ComponentWithJobID.contextTypes = {
      reactJobsClient: PropTypes.object.isRequired,
      reactJobsServer: PropTypes.object,
    };
    return ComponentWithJobID;
  }

  // We wrap the standard job implementation with our SSR behaviour.
  return withSSRBehaviour(clientOnlyJob(work)(Component));
};

export default job;
