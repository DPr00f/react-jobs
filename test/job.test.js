/* @flow */

import React from 'react';
import { shallow, mount } from 'enzyme';
import { warningsAsErrors } from './__helpers__';

// Under Test.
import job from '../src/job';

const resolveAfter = (time : number, result : any = null) => new Promise(resolve =>
  setTimeout(() => resolve(result), time),
);
const rejectAfter = (time : number, error : any) => new Promise((resolve, reject) => {
  setTimeout(() => reject(error || new Error('poop')), time);
});
const validOptions = { work: () => resolveAfter(1) };
const workTime = 10; // ms
const Foo = () => <div>Foo</div>;

describe('job()', () => {
  warningsAsErrors();

  describe('options', () => {
    it('returns a function', () => {
      const actual = typeof job({ work: () => undefined });
      const expected = 'function';
      expect(actual).toEqual(expected);
    });

    it('should throw if no options are provided', () => {
      // $FlowIgnore: we expect this to flow error
      expect(() => job()).toThrowError(/must provide an options object/);
    });

    it('should throw if invalid type provided as options', () => {
      // $FlowIgnore: we expect this to flow error
      expect(() => job(1)).toThrowError(/must provide an options object/);
    });

    it('should throws if no work is provided', () => {
      // $FlowIgnore: we expect this to flow error
      expect(() => job({})).toThrowError(/must provide a work function/);
    });

    it('should throws if the work is invalid', () => {
      // $FlowIgnore: we expect this to flow error
      expect(() => job({ work: 1 })).toThrowError(/must provide a work function/);
    });
  });

  describe('higher order component', () => {
    const hoc = job(validOptions);
    const actual = hoc(Foo);

    it('should return a "class" component', () => {
      // You can do the below to check for a class component.
      // @see http://bit.ly/2hSf2be
      expect(typeof actual).toEqual('function');
      expect(actual.prototype).toBeDefined();
      // $FlowIgnore
      expect(actual.prototype.isReactComponent).toBeTruthy();
    });

    it('should set the expected displayName', () => {
      expect(actual.displayName).toEqual('FooWithJob');
    });
  });

  describe('rendering', () => {
    it('throws if the work does not return a promise', () => {
      const InvalidFooWithJob = job({ work: () => () => undefined })(Foo);
      expect(() => shallow(<InvalidFooWithJob />))
        .toThrowError(/work\(props\) should return a Promise/);
    });

    it('should set "loading" when processing work', () => {
      const FooWithJob = job({ work: () => resolveAfter(workTime) })(Foo);
      const actual = shallow(<FooWithJob />).find(Foo).props();
      const expected = { job: { loading: true, result: undefined, error: undefined } };
      expect(actual).toMatchObject(expected);
    });

    it('should set "result" when work completes successfully', () => {
      const FooWithJob = job({ work: () => resolveAfter(workTime, 'result') })(Foo);
      const renderWrapper = mount(<FooWithJob />);
      // Allow enough time for work to complete
      return resolveAfter(workTime + 5).then(() => {
        const actual = renderWrapper.find(Foo).props();
        const expected = { job: { loading: false, result: 'result', error: undefined } };
        expect(actual).toMatchObject(expected);
      });
    });

    it('should set "error" when work fails', () => {
      const FooWithJob = job({ work: () => rejectAfter(workTime, 'error') })(Foo);
      const renderWrapper = mount(<FooWithJob />);
      // Allow enough time for work to complete
      return resolveAfter(workTime + 5).then(() => {
        const actual = renderWrapper.find(Foo).props();
        const expected = { job: { loading: false, result: undefined, error: 'error' } };
        expect(actual).toMatchObject(expected);
      });
    });
  });
});
