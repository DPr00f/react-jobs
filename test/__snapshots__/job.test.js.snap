exports[`job() rendering should set the "result" immediately if the work does not return a promise 1`] = `
<FooWithJob>
  <Foo
    job={
      Object {
        "error": undefined,
        "inProgress": false,
        "result": "bob",
      }
    }>
    <div>
      bob
    </div>
  </Foo>
</FooWithJob>
`;
